### PR TITLE
Riccardo aldrigo patch 1

### LINKED ISSUE #29: Reduce computational cost by eliminating redundant for loop
Inside functions "calculate_p_bar" and "calculate_smap" the program respectively groups and compare the atomistic configurations with the CG configurations.
Specifically, in the second function it uses a computationally heavy for loop that takes one by one the atomistic configurations and looks for their CGed version inside "p_bar" in order to calculate the single contributions to the mapping entropy.

I propose it might be useful to exploit the grouping previously done with "calculate_p_bar": 
instead of only clustering together the atomistic configurations and saving  the number of how many of them end up in a single CG configuration (i.e. calculating Omega_1), also the indices of the clusterized configurations could be saved in one column of "p_bar". This way, when calculating the mapping entropy, we would only need to do a for loop on the CG configurations and on the saved indices.
This should significantly reduce the computational cost, in particular when the number of CG configurations is exponentially lower than the number of atomistic configurations.

### PR DESCRIPTION
**Pull request by RiccardoAldrigo**
Closes #29

### CODE REVIEW

**Comment by mgiulini** [comment]
Hi @RiccardoAldrigo, welcome to pymap, and thank you very much for contributing!

Your idea is very useful, I had in mind of improving the performances of the calculation of the mapping entropy since a long time, but time is always an issue :)

Going to the technicalities: I still haven't updated the README, but I have introduced some tox checks. In particular, if you run the following commands:
```
tox -e py310
tox -e lint
```
you can see a nice report of the test coverage and of the lint. Your code does not pass the lint, which is understandable (you didn't know this check existed eheh). Can you try to adjust it? Should be very easy following the suggestions!

Second point: I tried to run the code using the parameters file `parameters/parameters_m1.dat`, and it gives a huge number of warnings of this kind:
```
PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  at_clust.insert(at_clust.columns.size, "omega_1", 0, True)
```

This seems a bit more substantial, what do you think?..I think I will try to move away from pandas for such calculations in the near future, it's too unstable and difficult to interpret/generalize. But in the meantime your PR could be merged (if we solve this minimal issues :) )

**Comment by mgiulini** [comment]
hey @RiccardoAldrigo, thanks for the update! and congratulations to your first github contribution!

As for the warnings, that's a pity...indeed issue #25 should take care of this, ensuring that all users have the same environment! by the way I run the code on m1 and no warnings show up!

As for the output, you're right, issue #31 should take care of substituting the old results file with the new ones, which use the new logic. This allows to print the mapping in a consistent order, so that if you run the code two times in a row, the result table should be exactly the same. Therefore, the last thing I might ask you is to re-introduce the call to `output_mappings`. After that, we can merge!

PS: I updated the description of the PR, so that issue #29 will be authomatically closed :)

**Comment by RiccardoAldrigo** [comment]
Hi @mgiulini, thanks a lot!
I'm glad that the warnings' problem has been solved.
If I understood correctly I just have to delate the older commands and un-comment "output_mapping" from pymap right?
If so, I just corrected it.

Let me know if I have to correct more things.

File: pymap.py [hunk 1]
```diff
@@ -78,39 +78,28 @@ def main():
...
+                p_bar = calculate_pbar_indices(at_clust, cg_clust,
+                                               df.shape[0],
+                                               mapping)
+                smap = calculate_smap_fast(mapping, pr, p_bar)
+                cg_mappings[key] = (len(mapping), mapping,
+                                    list(at_clust.columns[mapping]),
+                                    hs,
+                                    hk,
+                                    smap,
+                                    smap_inf)
```
Comment (mgiulini) [inline pymap.py:94]: in terms of readability the previous version is better, with just one variable per line

File: pymap.py [hunk 2]
```diff
@@ -78,39 +78,28 @@ def main():
...
-                cg_mappings[key] = (
-                    len(mapping),
-                    mapping,
-                    list(at_clust.columns[mapping]),
-                    hs,
-                    hk,
-                    smap,
-                    smap_inf
-                )
+                smap_inf = calculate_smap_inf(n_at, ncg,
```
Comment (mgiulini) [inline pymap.py:81]: once you put the arguments in different lines, it's always better to have one argument per line, it's much easier for handling the code later!

**Review by mgiulini — APPROVED** [review]

**Comment by mgiulini** [comment]
hi @RiccardoAldrigo, I just added a few comment on the style of the calls. After that you can safely merge the PR!

Thanks again for contributing, feel free to think and design other improvements!

**Comment by mgiulini** [comment]
Hi @RiccardoAldrigo, no problem, I can do it :) it takes some time to realise how it works, but trust me, it's a very efficient way to collaborate and maintain codes!

**Pull request merged**